src/Autofac/Core/Resolving/Middleware/CircularDependencyDetectorMiddleware.cs

```diff
@@ -0,0 +1,113 @@
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using Autofac.Core.Resolving.Pipeline;
+
+namespace Autofac.Core.Resolving.Middleware
+{
+    /// <summary>
+    /// Common stage. Added to the start of all component pipelines.
+    /// </summary>
+    internal class CircularDependencyDetectorMiddleware : IResolveMiddleware
+    {
+        public const int DefaultMaxResolveDepth = 50;
+
+        public static CircularDependencyDetectorMiddleware Default { get; } = new CircularDependencyDetectorMiddleware(DefaultMaxResolveDepth);
+
+        private readonly int _maxResolveDepth;
+
+        public CircularDependencyDetectorMiddleware(int maxResolveDepth)
+        {
+            _maxResolveDepth = maxResolveDepth;
+        }
+
+        public PipelinePhase Phase => PipelinePhase.RequestStart;
+
+        public void Execute(ResolveRequestContextBase context, Action<ResolveRequestContextBase> next)
+        {
+            var activationDepth = context.Operation.RequestDepth;
+
+            if (activationDepth > _maxResolveDepth)
+            {
+                throw new DependencyResolutionException(string.Format(CultureInfo.CurrentCulture, CircularDependencyDetectorMessages.MaxDepthExceeded, context.Service));
+            }
+
+            var requestStack = context.Operation.RequestStack;
+
+            // The first one is the current resolve request.
+            // Do our circular dependency check.
+            if (activationDepth > 1)
+            {
+                var registration = context.Registration;
+
+                // Only check the stack for shared components.
+                foreach (var requestEntry in requestStack)
+                {
+                    if (requestEntry.Registration == registration)
+                    {
+                        throw new DependencyResolutionException(string.Format(
+                            CultureInfo.CurrentCulture,
+                            CircularDependencyDetectorMessages.CircularDependency,
+                            CreateDependencyGraphTo(registration, requestStack)));
+                    }
+                }
+            }
+
+            requestStack.Push(context);
```

Comment:
In a tight loop, I see `PushWithResize` shows up as 2.48% of the time. Should we initialize the stack with some predefined capacity to save on that? Would it impact the memory usage in an adverse way? I guess it depends on whether memory or perf is more important. Possibly a micro-optimization - 42ms/1700ms of work in this profiler run I did.

Reply:
Pre-defining the capacity would up the memory usage, but perhaps not by enough to outweigh the speed benefits? Worth profiling.

Do you know how large the stack got in your test case?

Reply:
I didn't track it. I basically took the ChildScopeResolveBenchmark, the DeepGraphResolveBenchmark, and the OpenGenericBenchmark and ran each one at a time and profiled. It was, IIRC, the OpenGenericBenchmark case that made me notice the stack size increase.

I'll note it down for later investigation.